### DISCUSSION TITLE
MSS_RESET_N_M2F from MSS takes over 2 seconds to be released after DEVICE_INIT_DONE from PFSOC_INIT_MONITOR is set

### DISCUSSION THREAD

**Discussion by henryding-emcore** [polarfire-soc/polarfire-soc-documentation#502] [General]
I'm using PolarFire SoC FPGA to implement my design. In MSS, I don't use DDR Memory. However, I measured on the scope and found that MSS_RESET_N_M2F from MSS takes over 2 seconds to be released after DEVICE_INIT_DONE from PFSOC_INIT_MONITOR is set. 
I have two questions: 
1. While MSS_RESET_N_M2F keeps active, does that mean MSS initialization is not done yet although DDR Memory is not used? 
2. Is there a way to get MSS_RESET_N_M2F to be released earlier? 
Note: since DDR Memory is UNUSED, I don't see any configuration for DDR Controller. 

Thanks in advance!

**Comment by hughbreslin** [comment]
Hey @henryding-emcore sorry for the delayed response.

The MSS_RESET_N_M2F is driven from an MSS register, you can set it at any point during the boot process, there is a function in the HAL which can be used for this and you can see its just setting a bit in a register - its [here](https://github.com/polarfire-soc/platform/blob/main/mpfs_hal/common/mss_peripherals.c#L155). The reason its being used in the way it is:
1. To stop anything in the fabric trying to access the MSS while its in reset
2. To give the MSS the ability to reset the fabric if needed

Its not mandatory to use and if you have logic that doesn't rely on the MSS and is independent there is no requirement to wait for it to assert, so you could either set it earlier in the boot process or not use it to drive a reset for anything independent. 

To answer your questions:
1. I'm not sure on the use case, if this is the HSS or bare metal - each will set this bit in different places in the examples, but primarily yes the reset is released after the MSS has come up and reached main.
2. I answered this with the explanation above but yes you can release it earlier by setting the appropriate register bit once you know its safe for your FPGA based logic to start up

**Comment by henryding-emcore** [comment]
Thanks for your response, Hugh!

My intention is to find out why MSS takes over 2 seconds to finish the initialization. I'm not sure if it's caused by user codes or factory codes. Is there any document talking about MSS initialization time and the way to short it? My design is bare metal with boot loader in Core 0.

Thanks,
Henry

**Reply by rockoguido** [reply]
Hi Henry,    Perhaps DDR training is slowing you down, this can be on the order of 2 seconds plus.    Are you using your own bootloader or one of ours from Microchip, we have 2 that we distribute as part of our bare metal examples for the PolarFire SoC, I may be off base, it's unclear your boot sequence that gets you to Main.

Regards,
Dave
